### PR TITLE
Add deep_walk function and tests

### PR DESCRIPTION
**Pull request by egibs**
### Background

Rather than overhauling the implementation of `deep_get` in https://github.com/panther-labs/panther-analysis/pull/804, this PR adds an entirely separate (though drop-in) replacement for `deep_get` called `deep_walk`.

`deep_walk` can traverse any arbitrary dictionary consisting of other dictionaries as well as lists (even deeply-nested lists) and return a value; empty lists are also handled safely.

### Changes

* Adds `deep_walk` to `panther_base_helpers.py`
* Adds test cases for `deep_walk` to `global_helpers_test.py`
    * Also adds generation code for test cases

### Testing

* All new tests pass as expected
* Two of the new test cases utilize randomly generated dictionary structures to test the value retrieval as well as cases where the default value should be returned
    * These test cases run 1,000 iterations each
    * While each of these iterations isn't "reproducible", the overall test case result for each is
* A manual test case is also provided to test specific edge cases (returning the last value for a duplicate key, nested empty lists, etc.)

### CODE REVIEW

File: rules/gcp_audit_rules/gcp_service_account_access_denied.py [hunk 1]
```diff
@@ -1,20 +1,16 @@
...
 
 
 def rule(event):
-    details = _get_details(event)
-    if len(details) > 0:
-        reason = deep_get(details[0], "reason", default="")
-        if reason == "IAM_PERMISSION_DENIED":
-            return True
-    return False
+    reason = deep_walk(event, "protoPayload", "status", "details", "reason", default="")
```
Comment (egibs) [inline rules/gcp_audit_rules/gcp_service_account_access_denied.py:12]:
```
GCP.Service.Account.Access.Denied
        [PASS] service-account.access-denied-should-alert
                [PASS] [rule] true
                [PASS] [title] [GCP]: [test-no-perms@test-project-123456.iam.gserviceaccount.com] performed multiple requests resulting in [IAM_PERMISSION_DENIED]
                [PASS] [dedup] [GCP]: [test-no-perms@test-project-123456.iam.gserviceaccount.com] performed multiple requests resulting in [IAM_PERMISSION_DENIED]
                [PASS] [alertContext] {"project": "test-project-123456", "principal": "test-no-perms@test-project-123456.iam.gserviceaccount.com", "caller_ip": "12.12.12.12", "methodName": "google.iam.admin.v1.CreateServiceAccount", "resourceName": "projects/test-project-123456", "serviceName": "iam.googleapis.com"}
        [PASS] service-account.access-grated-should-not-alert
                [PASS] [rule] false
```

File: global_helpers/global_helpers_test.py [hunk 2]
```diff
@@ -1233,6 +1230,206 @@ def test_deep_get(self):
...
+        event = {
+            "key": {
+                "inner_key": [{"nested_key": "nested_value"}, {"nested_key": "nested_value2"}],
+                "very_nested": [
+                    {
+                        "outer_key": [
+                            {
+                                "nested_key": "value",
+                                "nested_key2": [{"nested_key3": "value2"}],
+                            }
```
[the remark this answers is not in the document]
Comment (egibs) [reply global_helpers/global_helpers_test.py]: Addressed in [`7d44a7b` (#819)](https://github.com/panther-labs/panther-analysis/pull/819/commits/7d44a7b428101410cc7a8867fe25f787196aaa44).

**Comment by egibs** [comment]
Currently working on the test case failure which is odd sense running that exact code works:
```
>>> deep_walk(event, "protoPayload", "status", "details", "reason")
'IAM_PERMISSION_DENIED'
```

File: global_helpers/global_helpers_test.py [hunk 3]
```diff
@@ -1254,6 +1252,221 @@ def test_deep_get(self):
...
+            "nested_value2",
+        )
+        self.assertEqual(
+            p_b_h.deep_walk(
+                event, "key", "inner_key", "nested_key", default="", return_val="first"
+            ),
+            "nested_value",
+        )
+        self.assertEqual(
+            p_b_h.deep_walk(event, "key", "very_nested", "outer_key", "nested_key", default=""),
```
Comment (edyesed) [inline global_helpers/global_helpers_test.py:1436]: NICE!

File: global_helpers/global_helpers_test.py [hunk 4]
```diff
@@ -1254,6 +1252,221 @@ def test_deep_get(self):
...
+            "value3",
+        )
+        self.assertEqual(p_b_h.deep_walk(event, "key", "another_key", default=""), "value6")
+        self.assertEqual(p_b_h.deep_walk(event, "key", "empty_list_key", default=""), "")
+        self.assertEqual(
+            p_b_h.deep_walk(event, "key", "empty_list_key", "nonexistent_key", default=""), ""
+        )
+        self.assertEqual(p_b_h.deep_walk(event, "key", "multiple_empty_lists_key1", default=""), "")
+        self.assertEqual(p_b_h.deep_walk(event, "key", "multiple_empty_lists_key2", default=""), "")
+        self.assertEqual(p_b_h.deep_walk(event, "key", "multiple_empty_lists_key3", default=""), "")
```
Comment (edyesed) [inline global_helpers/global_helpers_test.py:1456]: oh, natch! also nice

File: global_helpers/panther_base_helpers.py [hunk 5]
```diff
@@ -309,6 +310,60 @@ def deep_get(dictionary: dict, *keys, default=None):
     return out
 
 
+# pylint: disable=R1260,R0911
+def deep_walk(obj: dict, *keys: str, default: str = None, return_val: str = "all") -> Optional[str]:
```
Comment (edyesed) [inline global_helpers/panther_base_helpers.py:314]: it seem to me  the type hint for deep_walk's return value should be something like `-> Optional[str, List[str]]`

File: global_helpers/panther_base_helpers.py [hunk 6]
```diff
@@ -309,6 +310,60 @@ def deep_get(dictionary: dict, *keys, default=None):
...
+    """Safely retrieve a value stored in complex dictionary structure
```
Comment (edyesed) [inline global_helpers/panther_base_helpers.py:315]: I love this explainer, and I would love it more if this were formatted like a pydoc ala [LookupTableHelpers.p_match](https://github.com/panther-labs/panther-analysis/blob/master/global_helpers/panther_lookuptable_helpers.py#L27-L34)

File: global_helpers/panther_base_helpers.py [hunk 7]
```diff
@@ -309,6 +310,60 @@ def deep_get(dictionary: dict, *keys, default=None):
     return out
 
 
+# pylint: disable=R1260,R0911
```
Comment (edyesed) [inline global_helpers/panther_base_helpers.py:313]:
```suggestion
# pylint: disable=too-complex,too-many-return-statements
```

**Review by edyesed — COMMENTED** [review]
This is awesome! LMK if you can bring in the documentation nits 👍

File: global_helpers/panther_base_helpers.py (hunk 6 above)
Comment (egibs) [reply global_helpers/panther_base_helpers.py:315]: Will do!

Comment (egibs) [reply global_helpers/panther_base_helpers.py:315]: Addressed in [`37575e8` (#819)](https://github.com/panther-labs/panther-analysis/pull/819/commits/37575e805c2c8a2cb3b7ad02c92bc00bc693fa42).

File: global_helpers/panther_base_helpers.py (hunk 7 above)
Comment (egibs) [reply global_helpers/panther_base_helpers.py]: Addressed in [`37575e8` (#819)](https://github.com/panther-labs/panther-analysis/pull/819/commits/37575e805c2c8a2cb3b7ad02c92bc00bc693fa42).

File: global_helpers/panther_base_helpers.py (hunk 5 above)
Comment (egibs) [reply global_helpers/panther_base_helpers.py]: I took it out but I agree that re-adding a more accurate type hint would be preferable.

Comment (egibs) [reply global_helpers/panther_base_helpers.py]: Addressed in [`d7b2e68` (#819)](https://github.com/panther-labs/panther-analysis/pull/819/commits/d7b2e68d4cf61894708d48c679908da41b9a5b00).

**Review by edyesed — APPROVED** [review]
adds `deep_walk`, which has the feature of being able to walk lists, too!
